### DISCUSSION TITLE
Welcome to NinjaTrader8 Discussions!

### DISCUSSION THREAD

**Discussion by MicroTrendsTom** [MicroTrendsLtd/NinjaTrader8#41] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Sidlercom** [comment]
Hello everyone, I am new to programming and would like to learn a lot about the great possibility ATSQuadroStrategyBase here.
In the ninjatrader forum i once had a discussion with NinjaTrader_Jim about a strategy in unmanaged mode and Rithmic datafeed.

[https://ninjatrader.com/support/forum/forum/ninjatrader-8/strategy-development/1086069-unmanaged-mode-with-partialfilled-orders-and-trailingstop/page2](url)

He wrote to me in response:
**Rithmic uses separate feeds for Order Execution and Position events. While NinjaTrader makes an effort to ensure that Executions happen after Order events, it cannot always ensure this ordering. This is a limitation of NinjaTrader when using Rithmic.** 

I was never able to create my strategy with Rithmic data feed back then without possible errors in partial filling. 
Now my question: Are there no problems with Rithmic, or how is the problem solved?
If I have understood all this correctly, everything is monitored and regulated via the TradeWorkFlow instead of via OnOrderUpdate() and OnExecutionUpdate()?
Because i had the problem with partial filling and rithmic, i tried to work with only OnOrderUpdate() and calculate my position() myself. But this is not a good solution either, because without OnExecutionUpdate() it is not guaranteed that all orders are really filled.

I would like to thank you for providing me with such an opportunity here!!! And I apologise for my poor English

**Reply by MicroTrendsTom** [reply]
I am not aware of an issues with Rithmic  + ATSQuadroStrategyBase  as yet.
In the OnOrderUpdate it uses some work arounds to get/set references to order objects and this happens pre order routing send - 
So OnExecution would follow after that.
Generally speaking Partial fills are ignored and filled order executions are processed.

Try it out and if you get any issues pls report in
You can just download the zip code and then copy it into NT8 and try out the very basic Strategies 
And then feedback if any issues...  you can try this on rithmic demo account so its not local sim101  that would be closest to live trading

**Reply by Sidlercom** [reply]
Hi Tom, I finally have a demo account with Rithmic data feed, and it has been confirmed here as well that there are problems with partial executions, as I have already described it above. The problem is that the Rithmic API does not call the OnOrderUpdate and OnExecutionUpdate in order. The following I found out: if a partial execution is done, according to OnExecutionUpdate is called according to OnOrderUpdate and if the complete order is filled there, there is no call to the OnExecutionUpdate, which prevents the TradeWorkFlow from being set to GoLongSubmitOrderFilled or GoShortSubmitOrderFilled. The position was then opened, but no stops and targets are set.

This is what Ninjatrader says:
Rithmic and Interactive Brokers Users: Due to provider API designs these adapters do not guarantee the sequence of events of OnOrderUpdate, OnExecution, and OnPositionUpdate. Therefore, when working on a strategy that will run on these connections it is best practice to only work with passed by value data from that callback to eliminate the dependency on the sequence of events.

[NinjaScript Ausgabe 13.05.2021 19_49.txt](https://github.com/MicroTrendsLtd/NinjaTrader8/files/6477497/NinjaScript.Ausgabe.13.05.2021.19_49.txt)

**Reply by Sidlercom** [reply]
My solution looks like this so far:

in OnOrderUpdate()
```
//Only required for Rithmic and InteractivBrokers data feed.
//TradeWorkFlow for partial executions from OnExecutionUpdate(), here confirm as soon as the order has been fully executed.
if (order.OrderState == OrderState.Filled && order == orderEntry && isOrderEntryPartialFilled && (orderEntry.OrderAction < OrderAction.Sell ? TradeWorkFlow < TradeWorkFlowState.GoLongSubmitOrderFilled : TradeWorkFlow < TradeWorkFlowState.GoShortSubmitOrderFilled))
   {
      isOrderEntryPartialFilled = false;
      TradeWorkFlow = orderEntry.OrderAction < OrderAction.Sell ? TradeWorkFlowState.GoLongSubmitOrderFilled : 
      TradeWorkFlowState.GoShortSubmitOrderFilled;
      ProcessWorkFlow();
   }
```
in OnExecutionUpdate()
```
 if (execution.Order == orderEntry && (execution.Order.OrderState == OrderState.PartFilled || execution.Order.OrderState == OrderState.Cancelled && execution.Order.Filled > 0))
   {
       isOrderEntryPartialFilled = true;
   }
 if (execution.Order == orderEntry && (orderEntry.OrderAction < OrderAction.Sell ? TradeWorkFlow < TradeWorkFlowState.GoLongSubmitOrderFilled : TradeWorkFlow < TradeWorkFlowState.GoShortSubmitOrderFilled))
   {
        isOrderEntryPartialFilled = false;
        TradeWorkFlow = orderEntry.OrderAction < OrderAction.Sell ? TradeWorkFlowState.GoLongSubmitOrderFilled : TradeWorkFlowState.GoShortSubmitOrderFilled;
        ProcessWorkFlow();
        return;
   }
```
so far this solution works, but there is perhaps a better, more elegant solution, then please tell me thank you ;-)

**Reply by MicroTrendsTom** [reply]
Ok that's making more sense to the problem/caveats etc. Workarounds required.
I will have to get one of those datafeeds and experiment myself at some point 
thats great debugging by the way much appreciated -
So the crux was this:

partiall fill  -  then due to broker centric rules - instead of neutral broker rules  we then miss out the final call
>>>it doesn't go on here because Rithmic no longer calls the OnExecutionUpdate()

Ok i think i can see your amendments go in a derived strategy class   SidiSystemBase 
And you can call down to base.OnExec... base.OnOrder  as needed.
sounds like we need to cater for the above in the base class itself?

I think its best to open an Issus case/bug for this  
https://github.com/MicroTrendsLtd/NinjaTrader8/issues/45

**Reply by Sidlercom** [reply]
Yes the changes are all in the base class, I only work with the base class without subclasses.
The problem is known to ninjatrader, no idea how many years it will take them to find a solution to it, until it is ready, we have to solve it ourselves. If you need information, you can write to me, I will be happy to help you if I can ;-)

**Reply by MicroTrendsTom** [reply]
Excellent - please see the issue for details on this #45

**Comment by Sidlercom** [comment]
since only market orders are used for entry, the problem cannot occur ;-) will there also be limit orders for entry in the future?

**Comment by MicroTrendsTom** [comment]
You can program any orders you like into the samples.
The profit targets are limit. so they provide a case example of how to add limit orders for entry you would need to consider time to live

In time there will be samples and examples with the full gamut of features in layers above  -trading blocks to build strategies etc
all you do is add the entry logic

**Reply by Sidlercom** [reply]
i am interested to see, how the limit entry orders and partial filling will be solved, especially with the data feed from rithmic and the associated problems with partial executions ;-)

**Reply by MicroTrendsTom** [reply]
That depends how you the rules are coded for the strategy - wait for full fill, or on partial fill trigger etc.
For example generally onExecution ignores partial fills and waits for a filled order

**Reply by Sidlercom** [reply]
i already know that, but i'm curious how the partial executions are managed in connection with the data feed rithmic, because unfortunately i couldn't program it myself without errors. the problem was that rithmic uses different servers for the responses, which causes problems with limit orders and partial executions. i'll let myself be surprised what else comes up here, maybe there will be a suitable example for me at some point.

**Reply by MicroTrendsTom** [reply]
ok gotcha

**Reply by MicroTrendsTom** [reply]
you can try to add this untested line of code for limit?

 public override Order SubmitLong(string signal)
//long
 double price = Low[0] - 2 * TickSize;
orderEntry = SubmitOrderUnmanaged(0, OrderAction.Buy, OrderType.Limit, 4, price, price, String.Empty, signal);


 public override Order SubmitShort(string signal)
//Short
 double price = High[0] + 2 * TickSize;
orderEntry = SubmitOrderUnmanaged(0, OrderAction.SellShort, OrderType.Limit, 4, price, price, String.Empty, signal);

**Reply by Sidlercom** [reply]
unfortunately i don't have a rithmic data feed to test at the moment. i tested with the nt demo data feed and bought 100 contracts. if not all contracts could be bought, an open quantity remains and no stoploss and profit target orders are placed in the market, but that is normal because no partially executed orders are handled in OnExecution.

![Screenshot_2](https://user-images.githubusercontent.com/62642679/105998321-ca52d680-60ac-11eb-9327-2ec0bb3a5558.jpg)
![Screenshot_3](https://user-images.githubusercontent.com/62642679/105998346-d048b780-60ac-11eb-8712-78290867e49f.jpg)

**Reply by MicroTrendsTom** [reply]
yep probably needs an option to partially place stops and exits or chase in the limit order etc to market

**Reply by Ki2zer0** [reply]
In the case of "Limits": how do you check and cancel unfilled working, pending Limit orders for your AlgoSystemBase ?
> cancel after EntryBar + x*Bars, or do you have a better approach ? 

THX

**Reply by MicroTrendsTom** [reply]
Cancel after - barsSinceEntry >  OrderCancelBarsTimeOut
A type of Fill or Kill approach on partial fill detected  - move the limit order past the bid or ask for a fill

**Comment by Ki2zer0** [comment]
Hi Tom!

Thank you very much for your NT8 Code! 
I coded many approaches with ninjascript but they ended up in many errors and unprotected positions, etc...

Now, i tested your AlgoSystemBase 2021.3.1.1 with NT8 (CQG demo data) && interactive brokers paperAccount:

+ Do you have experiences and recommended settings ( in the Zystem params) for a clean run, and is "Realtime Trading 
    SignalQueue" here an option ?

+ I adapted your code in the Trademanagement #region to close the trade: try{ .....

if (IsLong && condition) { base.TradeWorkFlow = base.ProcessWorkFlow(StrategyTradeWorkFlowState.ExitTrade); AlgoSignalAction = AlgoSignalAction.ExitTradeLong;}

Is this the workflow to exit the whole trade (and ocos) or would you recommend another code ?

have a nice day,

christian

**Reply by MicroTrendsTom** [reply]
ok nice to hear that.  
TBH i didnt test much the SignalQ as of yet....  The premise is to avoid bumps of a fast market
However the engine in default mode will q signals and then execute them  or purge and deal with the most recent one
As long as its doing what you want in realtime no q mode needed but you can experiment

Exits
Exits can be a generic Exit  regardless of direction

**Reply by Ki2zer0** [reply]
with my setup (above) i have sometimes orphan trades with no stops and targets, on CQG-demodatafeed everything is ok....? No stops, sometimes with no error-messages, and partially failures like: "could not close pending OCO order". 
I experimented with all different settings!

> So what is your favor Real - data/execution-feed with your algos?

**Comment by Sidlercom** [comment]
Hi Tom, i have developed my own charttrader, which provides more information than the one from NT8. unfortunately, there seem to be performance problems with rendering or elsewhere, as the chart often freezes or even crashes NT completely. i am absolutely no programmer and have little experience. i have put everything together laboriously and with a lot of time. i don't know whether you or someone else here would be interested in taking a look at the project i have started, or even working on it to eliminate the existing problems or implement them better?
![Screenshot_1](https://user-images.githubusercontent.com/62642679/112620249-8e2bc000-8e28-11eb-9efe-dbfb98fde046.jpg)

**Reply by MicroTrendsTom** [reply]
Sorry no time - sounds like a locking issue  or data lagging handling issue or both combined - you could probably create a  github repos and then see if others are interested to collaborate etc

**Reply by pjsgsy** [reply]
" i don't know whether you or someone else here would be interested in taking a look at the project i have started, or even working on it to eliminate the existing problems or implement them better?" 

Give me a shout off-list to discuss if you like. We might be able to help each other. I have already completed something very similar over the last few years and solved many of these issues. pjs dot guernsey at gmail dot com.

**Reply by Sidlercom** [reply]
Hi @pjsgsy, thank you for your info, that sounds very good to me.  

I put the project on Github, but in german, because my english is not very good ;-)
I'm a complete novice at programming and also using GitHub, so any creative suggestions for improvement on your part would be greatly appreciated!

https://github.com/Sidlercom/SidiStrategyBaseNT8.git

if you like, you can download the files and have a look at them, i have made them freely available. i would be happy to hear from you on github and we could exchange information about our projects there.

**Reply by MicroTrendsTom** [reply]
Note: i will also release the GUI i wrote for this some time soon 
plus the layers and examples of how it all wires up etc

**Reply by Sidlercom** [reply]
> Note: i will also release the GUI i wrote for this some time soon
> plus the layers and examples of how it all wires up etc

I am really looking forward to this, it will certainly help me a lot in my project, thank you very much for your support!

**Reply by MicroTrendsTom** [reply]
please see the latest here if you didnt already -  you would need to remove the old files first i think...
https://github.com/MicroTrendsLtd/NinjaTrader8/discussions/61#discussion-3889698

**Comment by Ki2zer0** [comment]
I have Close-Errors in the WorkFlow, what does this mean? any ideas to fix this ? THX :)

2021.03.26 10:04:02:547 Strategy X Realtime Short Realtime ExitTradeClosePositionsPending:>OnOrderUpdate(↑Close#2 OrderId=1724021522 State=Filled)

2021.03.26 10:04:02:570 Strategy X Realtime Short Realtime ErrorFlattenAllPending:>  

after 6 Min.

2021.03.26 10:10:33:755 Strategy X Realtime Short Realtime ErrorFlattenAllConfirmed:>

and after 5 Min. the strat was ready to trade further:

2021.03.26 10:15:25:766 Strategy X Realtime Flat Realtime Waiting:>  


>> The strat closed at 10:04:02:547 the last trade, no orders left (on the chart) but the workFlow did not match and blocked new trade-entries!?

THX,

christian

**Reply by MicroTrendsTom** [reply]
It means the system took actions to resolve an error and closed the positions and orders 
This can be due to too fast a timeframe or a slow pc or NY8 got bogged down and some orders were not cancelled in time or a position was not closed in time so the workflow too action to close and cancel etc

if this keep recurring Please see here
https://github.com/MicroTrendsLtd/NinjaTrader8/issues/36

**Comment by efitzkiwi** [comment]
Hello, I have just started working with your engine and have a question about exiting a position.

My strategy went long in OnBarUpdate via 
`
if (!IsLong && some_buy_condition)
AlgoSignalAction = AlgoSignalAction.GoLong;
base.OnBarUpdate();
`
This also triggered my override functions for SubmitProfitTarget and SubmitStopLoss - which is fine... ALL GOOD ;)

My problem is, if I had some condition met in OnBarUpdate() that should CLOSE/EXIT a long trade AND its OCO orders... I need to do this with a function call, because after the CLOSE/EXIT of a long I need to go SHORT on the same bar. The problem is, 
`
AlgoSignalAction = AlgoSignalAction.GoLong;  
base.OnBarUpdate();
`
Will process the FINAL value of AlgoSignalAction ... what if I need to CLOSE/EXIT and then submit a SHORT on the same OnBarUpdate() call? Would that look like this?:
`
if (IsLong && some_sell_condition)  
{  
base.TradeWorkFlow = base.ProcessWorkFlow(StrategyTradeWorkFlowState.ExitTrade);  
AlgoSignalAction = AlgoSignalAction.GoShort;  
base.OnBarUpdate();  
}
`

Is this assessment correct for my situation?

**Reply by MicroTrendsTom** [reply]
GoLong will close any short position and enter a long
GoShort will close any long position and enter a short
ExitTrade will exit and do nothing else

the samples provided should demonstrate this behaviour

**Comment by TreyThomasCodes** [comment]
I just wanted to say how happy and excited I am to have found this project! There’s no telling how many hours of my free time that I spent over the last few years first trying to make my strategies conform to the managed order system, then trying to implement an unmanaged system that actually worked… 

Having now read through this code, I can’t even estimate what it would have taken me to produce something as robust and stable as this, if I **ever** would have. With two afternoons of work, I have my favorite strategy implemented and working correctly in a Contunuum demo account. I will still have to solve the well-documented Rithmic issues before it can be run in anger against the real market, but to be in the position of solving minutiae like that after only a handful of hours of work is truly remarkable. 

I can’t wait to see the improvements that will hopefully come to this through your commercial NT8 algo work, but even if nothing further comes out, this is still an incredible gift to the NT8 community that more people should know about.

**Reply by MicroTrendsTom** [reply]
Awesome you got it and you get it..  please share in the forums or wherever you can so more people can benefit.
I am very much wanting to return to this to build on this in layers - others are welcome to build on it too- but currently i'm tied up and cant come up for air until some other tech is done.  glad you are liking it - makes it worthwhile to get the feedback.

**Comment by lavalampmj** [comment]
I just discovered this framework today and have learned alot about C#, events, state machines and locks. Thanks for the contribution. I'm trying to get my head around when it makes sense to use this unmanaged framework as opposed to the NT managed approach?  Beyond breakout type systems, where opposing OCO stop entry long and short stop market orders can be placed (which is something sorely lacking in NT's managed approach).  What other style of systems can be implemented with this that the managed approach does not support?  

Does the OCO entry support stradling current market price with two orders in the same direction? e.g. LongLimit and Long Stop Market orders above and below the current price? 

Again, thanks for the contribution I really hope to use it.

**Reply by MicroTrendsTom** [reply]
About brief details on the approach here in the readme: https://github.com/MicroTrendsLtd/NinjaTrader8  end section.

NT default layer is not the most reliable or flexible etc.. but is a great start and suits traders in attendance or beginners to intermediate trader/devs.
Over time by the process of discovery an experienced developer trader will know all the pitfalls and limitations and prefer managed mode as it is incredibly powerful and the best system approach for retail trading beating all other platforms hands down and commerical apis savings $1000s oer month in api fees.

This engine just saves a lot of time for devs who want that mode
So its aimed at experienced developers and trader/dev's etc
But... ATSQuadroStrategyBase  engine is limited as per read me 4 brackets etc non compounding
but offers far superior trading and reliability in the unmanaged mode
Some sample code is provided for some ideas
Allows hybrid mode also interaction in realtime

OCO entry is long and short - you can program  a different way with overriding and editing your own code version etc
you can also open it up and make it do more what you need as its open source etc


hope that helps sorry super busy :-)

**Reply by lavalampmj** [reply]
After using the framework for a couple of days I can say it's really very excellent!  There is one rough spot that need to be smoothed out which may require changes to NT's core in order to help make this an unattended trading engine solution and also a couple of nice to have features.

Rough spot:

with:

IsRaiseErrorOnAllOrderRejects = false;
RealtimeErrorHandling = RealtimeErrorHandling.IgnoreAllErrors;

NT's broker sim throws an order error and results in UI pop up,  for example while attempting to place an order with a then invalid price. A situation that can happen all to regularly in fast volatile markets with non market orders. The strategy seems to hang with logic suspended until the "OK" is clicked, after which things get very untidy.  I see some bools declared but not used anywhere, I guessing this issue has been seen, thought about but not yet solved?

        protected bool onErrorShowAlertOnBarUpdate = false;
        protected bool onErrorDisableOnBarUpdate = false;
        protected bool onErrorShowAlertOnTerminate = false;

Recovering from these sorts of errors should be under the framework's or strategies programmatic control and not require user input.  I'll check the NT forums, I'm sure this must be a known issue and maybe requests outstanding?

Nice to haves:

1) Generic OCO entry order support. 

Currently, there is one entry OCO style explicitly contemplated, that being using opposing long and short oco stop market entry orders.
Making is clearer that the framework does actually support other OCO entry orders styles (a.k.a. both longs or both shorts that can be used to reverse an existing position) by making the method and const naming generic it would lower the newbie learning curve a tad.

2) ThouShallNotEnter() 

Obviously wrapping your order entry logic with time-based  filters is easy to do, the first idea I had was to offer the user a property so that they can define that no entries can occur N seconds or minutes before close of any session.   On further thought, the requirement might be a little bit more generic that just time based, for example - perhaps the system is not designed for ultra high volatility periods,  it would be great to have a master switch that stops any new entry orders.

If there is solution to the ui pop ups, and therefore a clear path to unattended operation, I'd definitely add the two above features for my own trading and if there interest submit via pull requests.

**Reply by MicroTrendsTom** [reply]
protected bool onErrorShowAlertOnBarUpdate = false;
protected bool onErrorDisableOnBarUpdate = false;
protected bool onErrorShowAlertOnTerminate = false;

OOPS! In fact these are a  redundant relic from the prior migrated source code base :-)
Somewhere here in discussions i guess... there was talk about a popup killer as a suggestion.... and NT forum may have it.
That would be fantastic if you look and see many thanks.

NOTE: Id say at nest NT8 strategies are pseudo unattended - if NT connection goes, it goes. internet and so on, 
if restart then can lose context...  some notes on others here:
https://microtrends.co/algorithmic-trading-systems/


Suggestions
i'm thinking we need to have a " nice to haves/suggestions" discussion thread for these items... and then look to implement etc

would certainly be into the idea of adding them to main

1) OCOC generics a good idea yes like it... 

2)  Yes nice...  Ddo not enter because a  pending exit time is a good filter to have - i use them in other systems.
generally i have an exit time trigger and period prior and after as a "no trade window"  filter 

So  This could be defined as Time Start and time End window  and there could be mulitple definitions
pre market  morning, lunch, close time etc

Crudely  example with 1 parameter to control which is on or off etc

Bitwise Flags
int FilterNoTradeTimeWindowMode = 1 | 2 | 4

mode 1
TimeSpan FilterNoTradeTimeWindowStart01 =  08:00
TimeSpan FilterNoTradeTimeWindowEnd01  =  09.29

mode 2
TimeSpan FilterNoTradeTimeWindowStart02 = 12.00
TimeSpan FilterNoTradeTimeWindowEnd02 = 13.30

mode 4
TimeSpan FilterNoTradeTimeWindowStart03 = 15:500
TimeSpan FilterNoTradeTimeWindowEnd03 = 18:00


//return true
bool ValidateFilterNoTradeTimeWindow(int filterNoTradeTimeWindowMode, TimeSpan start, TimeSpan end)
{
bool isNotInWindow
 etc

return isNotInWindow;
}
